### PR TITLE
Fwd gemm invokers

### LINKED ISSUE #519: Fwd gemm is crashing with int8 and group_count > 1
Here are 3 driver commands:
```Powershell
> bin/MIOpenDriver convint8 -F 1 -x 1 -y 1 -c 8 -g 1 #int8, group_count == 1
> bin/MIOpenDriver conv -F 1 -x 1 -y 1 -c 8 -g 4 #float, group_count > 1
> bin/MIOpenDriver convint8 -F 1 -x 1 -y 1 -c 8 -g 4 #int8, group_count > 1
```

The first and second run just fine. The third one crashes on the run stage. I have found this behavior because I have actually found the code responsible for that.

On the find stage, we have an `int8` check on the outside with no `group_count` check in the `int8` case.
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/5edf26d95e3840b21e9137e08d5e0661f91ad851/src/ocl/convolutionocl.cpp#L499

On the run stage, we have a `group_count` check on the outside with no `int8` check in the `group_count > 1` case.
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/5edf26d95e3840b21e9137e08d5e0661f91ad851/src/ocl/convolutionocl.cpp#L1129

So for `int8 && group_count > 1` case we have an evaluation inconsistent with run. Moreover, run treats it as a `float` case, I think.

It is fixed in #473.

Here are the logs for the commands:
[g1log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5392249/g1log.txt) - int8, group_count == 1, works fine
[fp_g4log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5392245/fp_g4log.txt) - float, group_count > 1, works fine
[g4log.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/5392250/g4log.txt) - int8, group_count > 1, crashes

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
- Find-db is updated with new solver IDs right here.
- By-product: Fixes #519 by making gemm not applicable to the case.

Forward gemm is now split between the following solvers:
- GemmFwd1x1_0_2 - applicable when 2d && fp32 && kernel is 1x1 && conv pads are 0x0 && conv strides are 2x2
- GemmFwd1x1_0_1_int8 - applicable when weights are int8 && kernel is 1x1 && conv pads are 0x0 && conv strides are 1x1 && groups_count is 1
- GemmFwd1x1_0_1 - applicable when weights are <b><i>not</i></b> int8 && kernel is 1x1 && conv pads are 0x0 && conv strides are 1x1
- GemmFwdRest - applicable when none of the solvers above is.

The fix of #519 is the part where GemmFwd1x1_0_1_int8 is not applicable with `group_count != 1` and GemmFwd1x1_0_1 is not applicable with int8.

"The rest" kind of the solver logic is flawed, but at the moment, it is implemented as it worked before. Changes to that may be applied in separate PR later.

### CODE REVIEW

**Comment by DrizztDoUrden** [comment]
I will have to split fwd gemm to more than one solver, so this is not as trivial as I thought.

**Comment by DrizztDoUrden** [comment]
I am not sure who to add to review gemm related code.

**Comment by atamazov** [comment]
@DrizztDoUrden Merge conflicts ;(

Please also update PR description with explanation of Solver ID changes, and inform @JehandadKhan as this means that Tuna needs to be updated.

I am beginning the testing round right now.

**Comment by atamazov** [comment]
@DrizztDoUrden 
> I am not sure who to add to review gemm related code.

@ce1adon I guess. @daniellowell will add more reviewers, if needed.

**Comment by DrizztDoUrden** [comment]
> inform JD as this means that Tuna needs to be updated.

Should I write @JehandadKhan an email or GitHub mention is sufficient?

**Comment by atamazov** [comment]
@DrizztDoUrden I believe he reads github, but nothing prevents us from duplicating in mail (for the sake of reliability, because this is an important topic). However, what we are going to tell is more important. I think we should:
- Give a clear warning that this PR breaks find-db generation from Tuna.
- Explain what changes need to be made there.
  - In particular, we need to clearly explain the applicability of the new Gemm solvers in the PR description (topmost comment).

/cc @JehandadKhan

**Comment by atamazov** [comment]
@DrizztDoUrden 
> - GemmFwd1x1_0_1 - applicable when ... conv strides are ___2x2___
> - GemmFwd1x1_0_2 - applicable when ... conv strides are ___1x1___

Is it really so?

**Comment by DrizztDoUrden** [comment]
No, typo.

**Comment by atamazov** [comment]
## :cyclone: Local performance & correctness testing

- Radeon VII, ROCm 3.5.
- develop at a6003098c vs this PR 10eca281f
- GPU time, Wall time, Aux Wall time were examined, as well as correctness.
- NORMAL Find mode, Immediate mode
  - HIP backend, 93 known FP32 + 93 known FP16 configs - :heavy_check_mark: _NO REGRESSIONS_
  - OCL backend, 93 known FP32 configs  - :heavy_check_mark: _NO REGRESSIONS_

**Review by atamazov — DISMISSED** [review]
LGTM!

**Comment by atamazov** [comment]
@DrizztDoUrden At `Hip clang embed build` step,  `test_conv_embed_db` has failed twice in a row. Please check locally.

**Comment by atamazov** [comment]
Oops, clang tidy error again:
image

**Comment by atamazov** [comment]
@DrizztDoUrden Merge conflicts ;(

**Comment by daniellowell** [comment]
@DrizztDoUrden Have we added enough flexibility in this to accommodate the potential change to MIOpenTensile?

**Comment by DrizztDoUrden** [comment]
@daniellowell Do you want to add another gemm solver, implementing some new tensile logic? If so, then yes, I have implemented the ability to add gemm forward solvers. If not, tensile changes would likely almost exclusively affect the gemm backend file (gemm.cpp or something).

**Comment by atamazov** [comment]
@DrizztDoUrden #709 merged. Please update this from develop.

**Comment by daniellowell** [comment]
This needs some reviews!

**Comment by atamazov** [comment]
@DrizztDoUrden Blocking issue:

- Build baseline library.
- Run one of the affected configs (see list below) in NORMAL Find Mode, so ufdb is updated
- Build library from PR
- Run the same config in immediate mode -> core dump.

I think this happens when old solver was `gemm`, but new one has different name, e.g. `GemmFwdRest`.

<details><summary><b><i>Example</i></b></summary>

```bash
atamazov@morpheus:~/github/MLOpen1/build/release.opencl$ ./bin/MIOpenDriver conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 8 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -S 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atam
azov/miopen-verify2 -S 0
MIOpenDriver conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 8 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0 -V 1 -s 0 -S 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2 -S 0
MIOpen(OpenCL): Info [Handle] Device name: gfx906
MIOpen(OpenCL): Info [Handle] stream: 0x55d8f8ede4a0, device_id: 0x55d8f8ebb230
MIOpen(OpenCL): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(OpenCL): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, MIOpen version 2.11.0.8416-2090356c8
MIOpen(OpenCL): Info2 [ValidateGcnAssemblerImpl] Running: '/opt/rocm/llvm/bin/clang --version'
MIOpen(OpenCL): Info2 [ValidateGcnAssemblerImpl] clang version 12.0.0 (/src/external/llvm-project/clang dac2bfceaa8d4a90257dc8a6d58f268e172ce00e)
MIOpen(OpenCL): Info2 [ValidateGcnAssemblerImpl] Target: x86_64-unknown-linux-gnu
MIOpen(OpenCL): Info2 [ValidateGcnAssemblerImpl] Thread model: posix
MIOpen(OpenCL): Info2 [ValidateGcnAssemblerImpl] InstalledDir: /opt/rocm/llvm/bin
MIOpen(OpenCL): Info2 [ValidateGcnAssemblerImpl]
MIOpen(OpenCL): Info2 [HipCompilerVersionImpl] Read version information from HIP package...
MIOpen(OpenCL): Info [HipCompilerVersionImpl] 4.0.20496
MIOpen(OpenCL): Info2 [AmdgcnAssembleQuiet] /opt/rocm/llvm/bin/clang  -x assembler -target amdgcn--amdhsa -mcpu=gfx900 /tmp/0e00-8228-f5ca-59d0 -o /dev/null 2>&1
Warm-up: Wall-clock Total Time: 491.783 ms, Find Algorithm: 1, Immediate Algorithm: miopenConvolutionAlgoDirect[13]
MIOpen(OpenCL): Info [GetForwardSolutionCount]
MIOpen(OpenCL): Info2 [FindRecordUnsafe] Looking for key 2048-7-7-1x1-512-7-7-8-3x3-2x2-1x1-0-NCHW-FP32-F in file /home/atamazov/github/MLOpen1/src/kernels/gfx906_60.OpenCL.2_11_0.ufdb.txt
MIOpen(OpenCL): Info2 [FindRecordUnsafe] Key match: 2048-7-7-1x1-512-7-7-8-3x3-2x2-1x1-0-NCHW-FP32-F
MIOpen(OpenCL): Info2 [FindRecordUnsafe] Contents found: miopenConvolutionFwdAlgoDirect:ConvDirectNaiveConvFwd,6.61855,0,miopenConvolutionFwdAlgoDirect,<unused>;miopenConvolutionFwdAlgoGEMM:gemm,0.70656,401408,MIOpenGEMM
,0_0_49_2048_49_49_512_2048;miopenConvolutionFwdAlgoWinograd:ConvBinWinogradRxSf2x3g1,1.7352,0,miopenConvolutionFwdAlgoWinograd,<unused>
MIOpen(OpenCL): Info2 [Measure] Db::FindRecord time: 0.084319 ms
MIOpen(OpenCL): Info [GetForwardSolutions]
MIOpen(OpenCL): Info2 [FindRecordUnsafe] Looking for key 2048-7-7-1x1-512-7-7-8-3x3-2x2-1x1-0-NCHW-FP32-F in file /home/atamazov/github/MLOpen1/src/kernels/gfx906_60.OpenCL.2_11_0.ufdb.txt
MIOpen(OpenCL): Info2 [FindRecordUnsafe] Key match: 2048-7-7-1x1-512-7-7-8-3x3-2x2-1x1-0-NCHW-FP32-F
MIOpen(OpenCL): Info2 [FindRecordUnsafe] Contents found: miopenConvolutionFwdAlgoDirect:ConvDirectNaiveConvFwd,6.61855,0,miopenConvolutionFwdAlgoDirect,<unused>;miopenConvolutionFwdAlgoGEMM:gemm,0.70656,401408,MIOpenGEMM
,0_0_49_2048_49_49_512_2048;miopenConvolutionFwdAlgoWinograd:ConvBinWinogradRxSf2x3g1,1.7352,0,miopenConvolutionFwdAlgoWinograd,<unused>
MIOpen(OpenCL): Info2 [Measure] Db::FindRecord time: 0.061826 ms
Forward Conv solutions available: 3
- id: 33 algo: 0, time: 0.70656 ms, ws: 401408, name: gemm
- id: 88 algo: 3, time: 1.7352 ms, ws: 0, name: ConvBinWinogradRxSf2x3g1
- id: 89 algo: 1, time: 6.61855 ms, ws: 0, name: ConvDirectNaiveConvFwd
MIOpen(OpenCL): Info [GetForwardSolutionWorkspaceSize] solver_id = gemm
Segmentation fault (core dumped)
$
$ # Now find mode to regenerate find-db
$ MIOPEN_LOG_LEVEL=4 ./bin/MIOpenDriver conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 8 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0 -V 0 -S -1
...
MIOpen Forward Conv. Algorithm: 0, Solution: 87/GemmFwdRest
...
$ # Immediate mode
$ MIOPEN_LOG_LEVEL=4 ./bin/MIOpenDriver conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 8 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0 -V 0 -S 0
...
- id: 87 algo: 0, time: 0.58496 ms, ws: 401408, name: GemmFwdRest
- id: 88 algo: 3, time: 1.73984 ms, ws: 0, name: ConvBinWinogradRxSf2x3g1
- id: 89 algo: 1, time: 6.54048 ms, ws: 0, name: ConvDirectNaiveConvFwd
```
</details>

<details><summary><b><i>Affected configs, OCL backend (FP16 not tested)</i></b></summary>

```
conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 8 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 16 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 512 -n 8 -k 256 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 1024 -n 8 -k 2048 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 512 -n 16 -k 1024 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 1024 -n 16 -k 512 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 1024 -n 16 -k 2048 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
```
</details>

<details><summary><b><i>Affected configs, HIP backend</i></b></summary>

```
conv -x 1 -y 1 -W 7 -H 7 -c 2048 -n 8 -k 512 -p 3 -q 3 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 112 -H 112 -c 64 -n 8 -k 64 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 56 -H 56 -c 64 -n 8 -k 256 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 128 -n 8 -k 512 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 512 -n 8 -k 256 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 256 -n 8 -k 1024 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 512 -n 8 -k 1024 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 1024 -n 8 -k 512 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 112 -H 112 -c 64 -n 16 -k 64 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 56 -H 56 -c 256 -n 16 -k 64 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 128 -n 16 -k 512 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 28 -H 28 -c 512 -n 16 -k 1024 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 1024 -n 16 -k 512 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
conv -x 1 -y 1 -W 14 -H 14 -c 1024 -n 16 -k 2048 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 56 -H 56 -c 64 -n 8 -k 256 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 56 -H 56 -c 256 -n 8 -k 64 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 28 -H 28 -c 128 -n 8 -k 512 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 28 -H 28 -c 512 -n 8 -k 128 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 28 -H 28 -c 512 -n 8 -k 256 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 14 -H 14 -c 256 -n 8 -k 1024 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 56 -H 56 -c 64 -n 16 -k 256 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 56 -H 56 -c 256 -n 16 -k 64 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 28 -H 28 -c 128 -n 16 -k 512 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 28 -H 28 -c 512 -n 16 -k 128 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 28 -H 28 -c 512 -n 16 -k 256 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 14 -H 14 -c 256 -n 16 -k 1024 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 14 -H 14 -c 1024 -n 16 -k 256 -p 0 -q 0 -u 1 -v 1 -l 1 -j 1 -g 1 -F 0
convfp16 -x 1 -y 1 -W 14 -H 14 -c 1024 -n 16 -k 2048 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 0
```
</details>

**Comment by atamazov** [comment]
@daniellowell I combine review & comprehensive testing

**Comment by atamazov** [comment]
Another issue, HIP backend, FP32 and FP16: The results of benchmarking the `GemmFwdRest` Solver differ than legacy `gemm` results. `GemmFwdRest` shows ~2..5 times _worse_ speed.

**Comment by DrizztDoUrden** [comment]
> Blocking issue

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/0c105b144655c2fd5f7e09a7de8480d7124461ad/CMakeLists.txt#L381

**Comment by DrizztDoUrden** [comment]
> Another issue, HIP backend, FP32 and FP16: The results of benchmarking of the GemmFwdRest solver during Find() are incorrect. GemmFwdRest shows ~2...5 times worse speed than it should be.

Have you checked if it happens on develop too? If not, I may check that.

**Comment by atamazov** [comment]
@DrizztDoUrden 

> > Another issue, HIP backend, FP32 and FP16: The results of benchmarking of the GemmFwdRest solver during Find() are incorrect. GemmFwdRest shows ~2...5 times worse speed than it should be.
> 
> Have you checked if it happens on develop too? ...

Yes, this is the essence of testing. I compare outputs of baseline (develop) with PR's outputs (logs, timings). Let's have a quick call tomorrow.

**Comment by atamazov** [comment]
@DrizztDoUrden 
> > Blocking issue
> 
> https://github.com/ROCmSoftwarePlatform/MIOpen/blob/0c105b144655c2fd5f7e09a7de8480d7124461ad/CMakeLists.txt#L381

This has been done to avoid performance degradation when using an unsuitable (e.g. outdated) find-db. However, an outdated database should not affect correctness (provided that database _format_ is correct).

IIRC when a non-applicable Solver is read from the find-db, it should be rejected, but the library dumps the core. _Something is wrong_, I guess.

Perhaps the issue is actually non-blocking, but we have to _prove_ that. I would say that this issue is non-blocking if it affects only outdated find-db in developer builds (i.e. "_something_" that is "_wrong_" does not affect the other aspects of behavior).

**Comment by atamazov** [comment]
@DrizztDoUrden I am sorry but something not good once more: ___FP16 perf drops.___

Example config:
```
./bin/MIOpenDriver convfp16 -x 1 -y 1 -W 14 -H 14 -c 256 -n 8 -k 1024 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -g 1 -F 1 -V 1 -s 0 -w 2 -t 1 -i 6 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2
```
<details><summary><b><i>Log, baseline</i></b></summary>

```
MIOpen(HIP): Info [get_device_name] Raw device name: gfx906
MIOpen(HIP): Info [Handle] stream: 0xe83c50, device_id: 0
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(HIP): Info [HipCompilerVersionImpl] 4.0.20496
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, MIOpen version 2.11.0.8351-2964633ef
Warm-up: Wall-clock Total Time: 3796.89 ms, Find Algorithm: 0
MIOpen(HIP): Info [ForwardGetWorkSpaceSize] 
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 1003520
MIOpen(HIP): Info [TryLoad] Find-db regenerating.
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoGEMM	0.069442	1003520
MIOpen(HIP): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: gemm , 1003520, 0.069442
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 1003520
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 1003520
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 1003520
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 1003520
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 1003520
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 1003520
Wall-clock Time Forward Conv. Elapsed: 0.193968 ms, Auxiliary API calls: 443.142 ms (GWSS: 0.0263672)
MIOpen Forward Conv. Algorithm: 0, Solution: 33/gemm
GPU Kernel Time Forward Conv. Elapsed: 0.084493 ms (average)
```
</details>

<details><summary><b><i>Log, PR</i></b></summary>

```
MIOpen(HIP): Info [get_device_name] Raw device name: gfx906
MIOpen(HIP): Info [Handle] stream: 0x1868c50, device_id: 0
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(HIP): Info [HipCompilerVersionImpl] 4.0.20496
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, MIOpen version 2.11.0.8435-ee71e5866
Warm-up: Wall-clock Total Time: 3818.4 ms, Find Algorithm: 0
MIOpen(HIP): Info [ForwardGetWorkSpaceSize] 
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 25088
MIOpen(HIP): Info [TryLoad] Find-db regenerating.
MIOpen(HIP): Info [FindSolutionImpl] GemmFwdRest (not searchable)
MIOpen(HIP): Info [EvaluateInvokers] GemmFwdRest: : 0.616728 < 3.40282e+38
MIOpen(HIP): Info [EvaluateInvokers] Selected: GemmFwdRest: : 0.616728, workspce_sz = 25088
MIOpen(HIP): Info [FindConvFwdAlgorithm] miopenConvolutionFwdAlgoGEMM	0.616728	25088
MIOpen(HIP): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: GemmFwdRest , 25088, 0.616728
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 25088
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 25088
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 25088
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 25088
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 25088
MIOpen(HIP): Info [ConvolutionForward] algo = 0, workspace = 25088
Wall-clock Time Forward Conv. Elapsed: 1.10638 ms, Auxiliary API calls: 345.527 ms (GWSS: 0.0390625)
MIOpen Forward Conv. Algorithm: 0, Solution: 87/GemmFwdRest
GPU Kernel Time Forward Conv. Elapsed: 0.643893 ms (average)
```
</details>

- _As you can see, workspace sizes are different._
- 16 out of total 93 configs are affected, average perf drop for affected configs is ~5X.
- ROCm 4.0, Radeon VII, environment:
```
export MIOPEN_FIND_MODE=normal
export MIOPEN_DEBUG_CONV_FFT=0
export MIOPEN_DEBUG_CONV_GEMM=1
export MIOPEN_DEBUG_CONV_WINOGRAD=0
export MIOPEN_DEBUG_CONV_DIRECT=0
export MIOPEN_DEBUG_CONV_IMPLICIT_GEMM=0
```
I am re-testing right now. The list of affected configs will be provided later.

**Comment by atamazov** [comment]
It seems like all 1x1 stride 2 configs are affected. The list of regressed configs:


No.of config in my   test set | W | H | c | n | k | x | y | p | q | u | v | Relative performance of Fwd Gemm -- PR vs baseline
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
89 | 14 | 14 | 1024 | 16 | 512 | 1 | 1 | 0 | 0 | 2 | 2 | 0.069702
54 | 14 | 14 | 256 | 16 | 1024 | 1 | 1 | 0 | 0 | 2 | 2 | 0.090972
75 | 14 | 14 | 1024 | 8 | 512 | 1 | 1 | 0 | 0 | 2 | 2 | 0.091948
45 | 14 | 14 | 256 | 8 | 1024 | 1 | 1 | 0 | 0 | 2 | 2 | 0.131236
85 | 28 | 28 | 512 | 16 | 256 | 1 | 1 | 0 | 0 | 2 | 2 | 0.137926
49 | 56 | 56 | 64 | 16 | 256 | 1 | 1 | 0 | 0 | 2 | 2 | 0.145944
51 | 28 | 28 | 128 | 16 | 512 | 1 | 1 | 0 | 0 | 2 | 2 | 0.147898
91 | 14 | 14 | 1024 | 16 | 2048 | 1 | 1 | 0 | 0 | 2 | 2 | 0.153049
82 | 56 | 56 | 256 | 16 | 128 | 1 | 1 | 0 | 0 | 2 | 2 | 0.154978
77 | 14 | 14 | 1024 | 8 | 2048 | 1 | 1 | 0 | 0 | 2 | 2 | 0.169856
42 | 28 | 28 | 128 | 8 | 512 | 1 | 1 | 0 | 0 | 2 | 2 | 0.207657
40 | 56 | 56 | 64 | 8 | 256 | 1 | 1 | 0 | 0 | 2 | 2 | 0.23131
68 | 56 | 56 | 256 | 8 | 128 | 1 | 1 | 0 | 0 | 2 | 2 | 0.241129
71 | 28 | 28 | 512 | 8 | 256 | 1 | 1 | 0 | 0 | 2 | 2 | 0.248041
87 | 28 | 28 | 512 | 16 | 1024 | 1 | 1 | 0 | 0 | 2 | 2 | 0.259134
73 | 28 | 28 | 512 | 8 | 1024 | 1 | 1 | 0 | 0 | 2 | 2 | 0.295325

**Comment by atamazov** [comment]
Very similar issue with BF16, same configs; just double-checked.

**Comment by atamazov** [comment]
# :tada: All performance tests passed!

File: src/solver.cpp [hunk 1]
```diff
@@ -381,6 +381,10 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
                        ++id,
                        ConvHipImplicitGemmWrwV4R4Xdlops_Padded_Gemm{},
                        miopenConvolutionAlgoImplicitGEMM);
+    RegisterWithSolver(registry, ++id, GemmFwd1x1_0_1{}, miopenConvolutionAlgoGEMM);
+    RegisterWithSolver(registry, ++id, GemmFwd1x1_0_1_int8{}, miopenConvolutionAlgoGEMM);
+    RegisterWithSolver(registry, ++id, GemmFwd1x1_0_2{}, miopenConvolutionAlgoGEMM);
+    RegisterWithSolver(registry, ++id, GemmFwdRest{}, miopenConvolutionAlgoGEMM);
```
Comment (atamazov) [inline src/solver.cpp:387]: I think this is merge error, but the new IDs must be ___appended___ at the end. 

Please also insert a clearly visible CAPSLOCK comment at the beginning and at the end of this code block.

**Review by atamazov — CHANGES_REQUESTED** [review]
Please fix solvers' registration. Otherwise fine.

File: src/solver/gemm.cpp [hunk 2]
```diff
@@ -0,0 +1,1080 @@
...
+#endif
+
+bool GemmFwdBase::IsApplicable(const ExecutionContext&,
+                               const conv::ProblemDescription& problem) const
+{
+#if MIOPEN_USE_GEMM
+    const auto& xDesc = problem.GetIn();
+    const auto& wDesc = problem.GetWeights();
+    const auto& yDesc = problem.GetOut();
+    return problem.GetDirection() == conv::Direction::Forward &&
```
Comment (atamazov) [inline src/solver/gemm.cpp:71]: Please also add check that layout is NCHW. This is for #744.

**Review by atamazov — CHANGES_REQUESTED** [review]
(CI is still not functional)

File: src/solver/gemm.cpp (hunk 2 above)
Comment (atamazov) [reply src/solver/gemm.cpp:71]: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/e0ffe166d7911ac6097c7d000e7a60ef126bd4f4/src/include/miopen/problem_description.hpp#L116

**Comment by atamazov** [comment]
Please fix tidy errors.

File: driver/conv_driver.hpp [hunk 3]
```diff
@@ -569,7 +569,7 @@ int ConvDriver<Tgpu, Tref>::AddCmdLineArgs()
 {
     inflags.AddInputFlag("in_layout", 'I', "NCHW", "Input Layout (Default=NCHW)", "string");
     inflags.AddInputFlag("out_layout", 'O', "NCHW", "Output Layout (Default=NCHW)", "string");
-    inflags.AddInputFlag("fil_layout", 'f', "NCHW", "Input Layout (Default=NCHW)", "string");
+    inflags.AddInputFlag("fil_layout", 'f', "NCHW", "Filter Layout (Default=NCHW)", "string");
```
Comment (atamazov) [inline driver/conv_driver.hpp:572]: Nice catch! :eye:

**Review by atamazov — APPROVED** [review]
LGTM!

**Comment by atamazov** [comment]
Converted to draft to avoid spoiling docker image cache on CI. You can continue work as usual. But please change this back to "normal" PR only after ending of CI moratorium (see mail). Then first merge from develop. Push the `[Ready for review]` button only after that. Thanks.

**Comment by atamazov** [comment]
@DrizztDoUrden The merge conflict is due to #777. I do not think you need to merge anything from develop to your version of convolutionocl.cpp. But you need to implement the workaround introduced in #777 in the respective Fwd GEMM solver's IsApplicable, otherwise CI test won't pass.

**Comment by atamazov** [comment]
Tidy error

**Comment by DrizztDoUrden** [comment]
Switched to mostly babysitting this one to finally get it merged.

**Review by atamazov — APPROVED** [review]

**Comment by atamazov** [comment]
@JehandadKhan This PR includes necessary find-db updates. Maybe you need to update Tuna databases, though.

**Comment by atamazov** [comment]
@DrizztDoUrden Now you can go to a disco or a bar :champagne:

**Comment by DrizztDoUrden** [comment]
Two more gemm directions are still here though.

**Pull request merged**